### PR TITLE
Property docs

### PR DESCRIPTION
**Pull request by karlbom**
## Summary

This PR is going to add API documentation as well es introductory documentation on how to interact and with and use `Properties`, `Property DDS` and `Property Binder`(aka `DataBinder`)

At the moment the API documentation lives under the experimental section in the docs section:

image

Additional Documentation and Guides I currently put in the getting started page in its own `Property` section.

image

I marked this PR as draft as the work on the documentation is still ongoing but it is probably a good idea to share the current state to discuss where these docs should be located and anything related to it.

There are some items still need to be done:
## Open Items
- [x] There are some changes I needed to do to the `custom-api-documenter` @tylerbutler provides a possible solution but there is still a bugfix that needs to go into the package (see https://github.com/karlbom/custom-api-documenter/commit/2918651ad07e949ba9e4059ae084285e1964be3b)
- [ ] final location of API docs
- [ ] final location of documentation and user guides etc.
- [x]  add Property DDS documentation
- [x]  add Property documentation (create, modify etc.)

Marking @SamBroner @DLehenbauer

### CODE REVIEW

**Comment by DLehenbauer** [comment]
@karmbom - FYI - Running `npm run policy-check:fix` from the root of the enlistment will sort the package.json file for you.
@tylerbutler - Thanks for helping out!

**Comment by karlbom** [comment]
@tylerbutler  I made a PR to your branch incorporating the changes I made to the custom-api-documenter  ( https://github.com/tylerbutler/custom-api-documenter/pull/1)

**Comment by tylerbutler** [comment]
@karlbom Thank you! I was hoping to switch to [this new fork](https://github.com/tylerbutler/rushstack/tree/custom-api-documenter) but when I tested yesterday the output wasn't what I expected, so I need to debug more. I don't want to hold things up, so I'll merge [your PR](https://github.com/tylerbutler/custom-api-documenter/pull/1) and update main to use it.

The fixes you made should already be included in my new fork, with the exception of the backwards-compatible schema change.

**Comment by karlbom** [comment]
Since the documentation turned out to be quite a lot I decided for the moment to split them into handling the basics of Property DDS and properties and the documentation around the Property DataBinder

image


@SamBroner you mentioned there is going to be some update around the documentation do you have any opinion on where the documentation for this PR should live or whether it is okay to be where it is for now?

**Comment by tylerbutler** [comment]
I'd like to propose doing one of two things:

1. Create a new section for DDSes and add the Property DDS docs there. OR
2. Create new empty sections under `docs` for all the DDSes, and add the Property DDS docs there.

Option 2 is probably the most expedient, though I worry about the nav being too complex. @SamBroner Did you have something else in mind?

**Comment by SamBroner** [comment]
> I'd like to propose doing one of two things:
> 
> 1. Create a new section for DDSes and add the Property DDS docs there. OR
> 2. Create new empty sections under `docs` for all the DDSes, and add the Property DDS docs there.
> 
> Option 2 is probably the most expedient, though I worry about the nav being too complex. Sam Broner Did you have something else in mind?

Let's start with 2, that is what I had in mind. I should have a PR for that tomorrow afternoon.

**Comment by tylerbutler** [comment]
Sounds good. Let me know if you need help.

**Comment by karlbom** [comment]
@SamBroner is there an update?

**Comment by karlbom** [comment]
@SamBroner thanks for the fixes

File: docs/content/start/property-dds/property-overview.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+rootProperty.insert('myProperty', stringProperty);
+
+console.log('Added one property to the root property');
+rootProperty.prettyPrint();
+```
+
+### What is Property Factory?
+
+The is a factory object for instantiating Properties. It comes with a set of rich primitive types (Float32, Enum, Uint16
+, Uint32, etc. and custom types) For a more detailed explanation please refer to the
```
Comment (kailui) [inline docs/content/start/property-dds/property-overview.md:64]:
```suggestion
, Uint32, etc. and custom types). For a more detailed explanation please refer to the
```

File: docs/content/start/property-dds/property-overview.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+`commit`the changes. This will be explained in more detail below.
+
+## Adding Data to Property DDS
+
+Data is added to the Property DDS via "commits"
+
+A Commit is the “unit of change” in Property DDS. It stores an atomic transaction recording every insert, modify and
+delete operations that happened since the previous commit.
+
+Much like in Git, a chain of commits is known as a “Branch”. The state of the is constructed by walking along the history and applying the commits on top of the previous one at any point in time. It is important to note that — since history cannot be modified — a must be referring the most recent state in order to persist any changes in Property DDS.
```
Comment (kailui) [inline docs/content/start/property-dds/property-overview.md:77]: Sounds like a word is missing here.

Comment (kailui) [inline docs/content/start/property-dds/property-overview.md:77]: > The state of the is constructed by

Sounds like a word is missing between "the" and "is".

Comment (kailui) [inline docs/content/start/property-dds/property-overview.md:77]: > a must be referring

Should there be a word after "a"?

File: docs/content/start/property-dds/property-overview.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+```javascript
+myPropertyDDS.commit()
+```
+
+## Read Data From The PropertySet
+
+Naturally, after writing data to a Fluid Property DDS we need the ability to retrieve this data. For this, we must initialize the Property DDS with the Fluid `documentId`from the Property DDS that we wrote or data into. Once the *Property DDS* is fully initialized by the FluidFramework you can retrieve that data.
```
Comment (kailui) [inline docs/content/start/property-dds/property-overview.md:85]:
```suggestion
Naturally, after writing data to a Fluid Property DDS we need the ability to retrieve this data. For this, we must initialize the Property DDS with the Fluid `documentId`from the Property DDS that we wrote our data into. Once the Property DDS is fully initialized by the Fluid Framework you can retrieve that data.
```

File: docs/content/start/property-dds/property-schemas.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
+---
+title: Schemas
+menuPosition: 2
+---
+
+A *Property Sets Schema* defines the structure of a Property Set. It defines which child properties exist in a
+property and which types they have.
+
+The following example shows a possible schema for your rectangle. This example is only meant to show the various possibilities in defining schemas and is not necessarily showing the best way to define a rectangle.
```
Comment (kailui) [inline docs/content/start/property-dds/property-schemas.md:9]:
```suggestion
The following example shows a possible schema for your rectangle. This example is only meant to show the various possibilities in defining schemas and is not necessarily the best way to define a rectangle.
```

File: docs/content/start/property-dds/property-schemas.md [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+```json
+{ "id": "position",
+  "properties": [
+    { "id": "x", "typeid": "Float64" },
+    { "id": "y", "typeid": "Float64" }
+  ]
+}
+```
+
+Or you can refer to an external schema, as we did for fillcolor.
```
Comment (kailui) [inline docs/content/start/property-dds/property-schemas.md:59]:
```suggestion
Or you can refer to an external schema, as we did for fill color.
```

File: docs/content/start/property-dds/property-schemas.md [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+**set**
+  The sub-property contains an unordered collection of NamedProperty.
+
+## length
+
+This is an optional keyword that can be used when context is ‘array’ to specify a fixed-length array. If no length is specified, the array length is dynamic.
+
+
+
+## nnotation
```
Comment (kailui) [inline docs/content/start/property-dds/property-schemas.md:146]:
```suggestion
## Annotation
```

File: docs/content/start/property-dds/property-schemas.md [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+```
+
+## Constants
+
+
+A schema can define read-only properties through Constants. They are declared in a “constants” array which is at the same level as the properties one.
+
+    Each constant should have an id, value and a valid typeid. The id must be unique per declared constants and properties, that is a constant and a property cannot have the same id.
+
+    A constant can be of primitive, enum, string, array, set, map, or custom type.
```
Comment (kailui) [inline docs/content/start/property-dds/property-schemas.md:231]:
```suggestion
A constant can be of primitive, enum, string, array, set, map, or custom type.
```

File: docs/content/start/property-dds/property-schemas.md [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+
+    They are read only properties; values and instances can be retrieved like any other properties.
```
Comment (kailui) [inline docs/content/start/property-dds/property-schemas.md:233]:
```suggestion
They are read-only properties; values and instances can be retrieved like any other properties.
```

File: docs/content/start/property-dds/property-schemas.md [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+    Constants can be overridden in inherited schemas. To do that, the inherited constant should keep the same 'id', 'typeid' and 'context' as the base one it overrides.
```
Comment (kailui) [inline docs/content/start/property-dds/property-schemas.md:234]:
```suggestion
Constants can be overridden in inherited schemas. To do that, the inherited constant should keep the same 'id', 'typeid' and 'context' as the base one it overrides.
```

File: docs/content/start/property-dds/property-primitive.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+
+
+
+|    Typeid     |  Description  |
+|---------------|---------------|
+| **Float32**   |   A 32 bit floating point value (according to IEEE 754) |
+| **Float64**   |  A 64 bit floating point value (according to IEEE 754) |
+| **Int8**      | A signed 8 bit integer value (with a valid range from −128 to 127) |
+| **Uint8**     | A unsigned 8 bit integer value (with a valid range from 0 to 255) |
```
Comment (kailui) [inline docs/content/start/property-dds/property-primitive.md:17]:
```suggestion
| **Uint8**     | An unsigned 8 bit integer value (with a valid range from 0 to 255) |
```

File: docs/content/start/property-dds/property-primitive.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+| **Int16**     | A signed 16 bit integer value (with a valid range from −32768 to 32767) |
+| **Uint16**    | A unsigned 16 bit integer value (with a valid range from 0 to 65536) |
```
Comment (kailui) [inline docs/content/start/property-dds/property-primitive.md:19]:
```suggestion
| **Uint16**    | An unsigned 16 bit integer value (with a valid range from 0 to 65536) |
```

Comment (kailui) [inline docs/content/start/property-dds/property-primitive.md:19]:
```suggestion
| **Uint16**    | An unsigned 16 bit integer value (with a valid range from 0 to 65536) |
```

File: docs/content/start/property-dds/property-primitive.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+| **Int32**     | A signed 32 bit integer value (with a valid range from −2,147,483,648 to 2,147,483,647) |
+| **Uint32**    | An unsigned 32 bit integer value (with a valid range from 0 to 4,294,967,295) |
+| **Int64**     | A signed 64 bit integer value (with a valid range from -9,223,372,036,854,775,808 to 9,223,372,036,854,775,807) |
+| **Uint64**    | An unsigned 64 bit integer value (with a valid range from 0 to 18,446,744,073,709,551,615) |
+| **Bool**      | A boolean value, either true or false |
+| **String**    | An unicode string |
```
Comment (kailui) [inline docs/content/start/property-dds/property-primitive.md:25]:
```suggestion
| **String**    | A Unicode string |
```

File: docs/content/start/property-dds/property-paths-references.md [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+root.insert('point', point);
+
+let map = PropertyFactory.create('shape:position2d-1.0.0', 'map');
+pointInMap = PropertyFactory.create('shape:position2d-1.0.0');
+map.insert('point2', pointInMap);
+root.insert('map', map);
+
+// Now we can resolve paths
+root.resolvePath('point.x');
+root.resolvePath('map[point].x');
```
Comment (kailui) [inline docs/content/start/property-dds/property-paths-references.md:35]:
```suggestion
root.resolvePath('map[point2].x');
```

File: docs/content/start/property-dds/property-paths-references.md [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+
+// Spaces don't need to be escaped
+root.resolvePath('with spaces.x');
+
+// Property DDS proxies a helper function to perform quoting where needed
+let {PathHelper} = require('@fluid-experimental/property-changeset');
+root.resolvePath(PropertyUtils.PathHelper.quotePathSegmentIfNeeded('"doubleQuotes"'));
+```
+
+A path that starts with a ``/`` it is an **absolute path**. This means the resolution will start at the root of the
```
Comment (kailui) [inline docs/content/start/property-dds/property-paths-references.md:89]:
```suggestion
A path that starts with a ``/`` is an **absolute path**. This means the resolution will start at the root of the
```

File: docs/content/start/property-dds/property-collection.md [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+menuPosition: 6
+---
+
+Property DDS supports three different types of collection Properties: ``array``, ``map`` and ``set``. Property DDS distinguishes between
+two different types of collections:
+
+**Primitive Collections**
+  Primitive collections are strictly typed with one of the Primitive Property Types (e.g. a String array or Float64
+  map). These collections directly contain the literal values of the given type. They do not support polymorphism, all
+  entries have to be of the same type The collection itself is a leaf in the PropertySet   tree. This means, there are
```
Comment (kailui) [inline docs/content/start/property-dds/property-collection.md:12]:
```suggestion
  entries have to be of the same type. The collection itself is a leaf in the PropertySet tree. This means, there are
```

File: docs/content/start/property-dds/property-overview.md (hunk 1 above)
Comment (DLehenbauer) [reply docs/content/start/property-dds/property-overview.md:64]: @kailui - Thank you for contributing proof reading fixes!

File: docs/content/start/property-dds/property-overview.md (hunk 2 above)
Comment (karlbom) [reply docs/content/start/property-dds/property-overview.md]: thanks I added the missing words in that section

**Comment by tylerbutler** [comment]
I'm adding an experimental section back to the docs in an upcoming PR. I'll include these docs in that section.

**Pull request closed without merging**